### PR TITLE
Enable io_uring

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

Allow MultiGet be faster for reads from disk

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: *(Please link PR)*
  - [ ] Backport for 3.10: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document: 
- [ ] oskar PR: https://github.com/arangodb/oskar/pull/542

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 1]
```diff
@@ -1491,6 +1498,20 @@ of outgrowing the maximum number of file descriptors the ArangoDB process
...
+          "--rocksdb.use-io_uring",
+          "By default enabled and use runtime check availaibility of io_uring. "
+          "So should be used with false only if some issues were detected.",
+          new BooleanParameter(&ioUringEnabled),
+          arangodb::options::makeFlags(
+              arangodb::options::Flags::Uncommon,
+              arangodb::options::Flags::DefaultNoComponents,
+              arangodb::options::Flags::OnAgent,
+              arangodb::options::Flags::OnDBServer,
+              arangodb::options::Flags::OnSingle))
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1512]:
```suggestion
          "--rocksdb.use-io_uring",
          "Check for existence of io_uring at startup and use it if available. "
          "Should be set to false only to opt out of using io_uring.",
          new BooleanParameter(&ioUringEnabled),
          arangodb::options::makeFlags(
              arangodb::options::Flags::Uncommon,
              arangodb::options::Flags::DefaultNoOs,
              arangodb::options::Flags::OsLinux,
              arangodb::options::Flags::DefaultNoComponents,
              arangodb::options::Flags::OnAgent,
              arangodb::options::Flags::OnDBServer,
              arangodb::options::Flags::OnSingle))
```

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 2]
```diff
@@ -1491,6 +1498,22 @@ of outgrowing the maximum number of file descriptors the ArangoDB process
...
+          "Should be set to false only to opt out of using io_uring.",
+          new BooleanParameter(&ioUringEnabled),
+          arangodb::options::makeFlags(
+              arangodb::options::Flags::Uncommon,
+              arangodb::options::Flags::DefaultNoOs,
+              arangodb::options::Flags::OsLinux,
+              arangodb::options::Flags::DefaultNoComponents,
+              arangodb::options::Flags::OnAgent,
+              arangodb::options::Flags::OnDBServer,
+              arangodb::options::Flags::OnSingle))
```
Comment (MBkkt) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:1514]:
```suggestion
              arangodb::options::Flags::Uncommon,
              arangodb::options::Flags::OsLinux,
              arangodb::options::Flags::OnAgent,
              arangodb::options::Flags::OnDBServer,
              arangodb::options::Flags::OnSingle))
```
fix

File: scripts/docker/build/build-manifest.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+docker manifest rm ${manifest}:${IMAGE_TAG}
+docker manifest rm ${manifest}:latest
+
+docker manifest create ${manifest}:latest \
+  --amend ${manifest}-latest-amd64 \
+  --amend ${manifest}:latest-arm64 \
+&& docker manifest create ${manifest}:${IMAGE_TAG} \
+  --amend ${manifest}:${IMAGE_TAG}-amd64 \
+  --amend ${manifest}:${IMAGE_TAG}-arm64 \
+|| echo "Error during docker multiarch manifest creation!" && exit 1
```
Comment (mpoeter) [inline scripts/docker/build/build-manifest.sh:17]: I don't think we should use `latest`, but really just `IMAGE_TAG`

File: scripts/docker/test/build-manifest.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+docker manifest rm ${manifest}:${IMAGE_TAG}
+docker manifest rm ${manifest}:latest
+
+docker manifest create ${manifest}:latest \
+  --amend ${manifest}-latest-amd64 \
+  --amend ${manifest}:latest-arm64 \
+&& docker manifest create ${manifest}:${IMAGE_TAG} \
+  --amend ${manifest}:${IMAGE_TAG}-amd64 \
+  --amend ${manifest}:${IMAGE_TAG}-arm64 \
+|| echo "Error during docker multiarch manifest creation!" && exit 1
```
Comment (mpoeter) [inline scripts/docker/test/build-manifest.sh:17]: I don't think we should use `latest`, but really just `IMAGE_TAG`

**Review by mpoeter — CHANGES_REQUESTED** [review]
I don't think we should use `latest` tag, but otherwise LGTM

File: scripts/docker/test/build-manifest.sh (hunk 4 above)
Comment (KVS85) [reply scripts/docker/test/build-manifest.sh]: Removed.

**Review by mpoeter — APPROVED** [review]

File: arangod/RocksDBEngine/RocksDBOptionFeature.cpp [hunk 5]
```diff
@@ -56,6 +56,13 @@
 #include <rocksdb/table.h>
 #include <rocksdb/utilities/transaction_db.h>
 
+// It's not atomic because it shouldn't change after initilization.
+// And initilization should happens before rocksdb initilization.
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBOptionFeature.cpp:60]:
```suggestion
// And initialization should happen before rocksdb initialization.
```

File: scripts/docker/README.md [hunk 6]
```diff
@@ -1,9 +1,14 @@
-We want images to be tagged based on the version of the base image as well was the versions of the most important components.
-For example `arangodb/build-alpine-x86_64:3.16-gcc11.2-openssl3.1.1`
-That image is based on Alpine 3.16 and has gcc 11.2 and OpenSSL 3.1.1 installed.
+We want images to be tagged based on the version of the base image as well was the versions of the most important components. In order to not break base branch please set random part with a branch-name dependency: i.e. `git rev-parse --short HEAD`.
```
Comment (jsteemann) [inline scripts/docker/README.md:1]:
```suggestion
We want images to be tagged based on the version of the base image as well as the versions of the most important components. In order to not break the base branch please add a random part with a branch-name dependency: i.e. `git rev-parse --short HEAD`.
```

File: scripts/docker/README.md [hunk 7]
```diff
@@ -1,9 +1,14 @@
...
 
-To automate and simplify this, for each dockerfile there is a build-image.sh scriptthat creates the image with the `latest`
-tag and afterwards automatically determines the relevant version numbers, tags the image accordingly, and prints the command
-to push the image.
+For example `arangodb/build-alpine:3.16-gcc11.2-openssl3.1.1-<RAND>-amd64`
+That image is based on Alpine 3.16 and has gcc 11.2 and OpenSSL 3.1.1 installed built for x86-64 architecture.
 
-After building and pushing the image, don't forget to update .circleci/base_config.yml so that the new image is actually used.
\ No newline at end of file
+To automate and simplify this, for each dockerfile there are build-image_*.sh scripts that creates an image with the `latest-<ARCH>` tag and afterwards automatically determines the relevant version numbers, tags the image accordingly, and prints the command to push the image.
```
Comment (jsteemann) [inline scripts/docker/README.md:6]:
```suggestion
To automate and simplify this, for each dockerfile there are build-image_*.sh scripts that create an image with the `latest-<ARCH>` tag and afterwards automatically determine the relevant version numbers, tags the image accordingly, and print the command to push the image.
```

File: scripts/docker/README.md [hunk 8]
```diff
@@ -1,9 +1,14 @@
...
+
+Each supported architecture (x86-64 and arm64) has own build-image_*.sh script.
```
Comment (jsteemann) [inline scripts/docker/README.md:8]:
```suggestion
Each supported architecture (x86-64 and arm64) has its own build-image_*.sh script.
```

File: scripts/docker/build/build-image_arm64.sh [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+GCC_VERSION=$(docker run --rm -it $image:latest-$arch g++ "--version" | grep -Po " \K\d+\.\d+(?=\.\d+ )")
+
+OPENSSL_VERSION=$(docker run --rm -it $image:latest-$arch sh -c "cat \${OPENSSL_ROOT_DIR}/lib/pkgconfig/openssl.pc" | grep -Po "Version: \K\d+\.\d+\.\d+")
+
+OS_VERSION=$(docker run --rm -it $image:latest-$arch cat /etc/alpine-release | grep -Po "\d+\.\d+")
+
+IMAGE_TAG=${OS_VERSION}-gcc${GCC_VERSION}-openssl${OPENSSL_VERSION}-$(git rev-parse --short HEAD)-$arch
+
+echo "Taging image as \"${IMAGE_TAG}\""
```
Comment (jsteemann) [inline scripts/docker/build/build-image_arm64.sh:16]:
```suggestion
echo "Tagging image as \"${IMAGE_TAG}\""
```

File: scripts/docker/build/build-image_x86-64.sh [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+GCC_VERSION=$(docker run --rm -it $image:latest-$arch g++ "--version" | grep -Po " \K\d+\.\d+(?=\.\d+ )")
+
+OPENSSL_VERSION=$(docker run --rm -it $image:latest-$arch sh -c "cat \${OPENSSL_ROOT_DIR}/lib64/pkgconfig/openssl.pc" | grep -Po "Version: \K\d+\.\d+\.\d+")
+
+OS_VERSION=$(docker run --rm -it $image:latest-$arch cat /etc/alpine-release | grep -Po "\d+\.\d+")
+
+IMAGE_TAG=${OS_VERSION}-gcc${GCC_VERSION}-openssl${OPENSSL_VERSION}-$(git rev-parse --short HEAD)-$arch
+
+echo "Taging image as \"${IMAGE_TAG}\""
```
Comment (jsteemann) [inline scripts/docker/build/build-image_x86-64.sh:16]:
```suggestion
echo "Tagging image as \"${IMAGE_TAG}\""
```

File: scripts/docker/test/build-image_arm64.sh [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+arch=arm64
+
+image=arangodb/test-ubuntu
+docker build  --platform linux/$arch -t $image:latest-$arch --file arangodb-test.Dockerfile .
+
+OS_VERSION=$(docker run --rm -it $image:latest-$arch cat etc/os-release | grep -Po "VERSION_ID=\"\K\d+\.\d+")
+
+IMAGE_TAG=${OS_VERSION}-$(git rev-parse --short HEAD)-$arch
+
+echo "Taging image as \"${IMAGE_TAG}\""
```
Comment (jsteemann) [inline scripts/docker/test/build-image_arm64.sh:12]:
```suggestion
echo "Tagging image as \"${IMAGE_TAG}\""
```

File: scripts/docker/test/build-image_x86-64.sh [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+arch=amd64
+
+image=arangodb/test-ubuntu
+docker build  --platform linux/$arch -t $image:latest-$arch --file arangodb-test.Dockerfile .
+
+OS_VERSION=$(docker run --rm -it $image:latest-$arch cat etc/os-release | grep -Po "VERSION_ID=\"\K\d+\.\d+")
+
+IMAGE_TAG=${OS_VERSION}-$(git rev-parse --short HEAD)-$arch
+
+echo "Taging image as \"${IMAGE_TAG}\""
```
Comment (jsteemann) [inline scripts/docker/test/build-image_x86-64.sh:12]:
```suggestion
echo "Tagging image as \"${IMAGE_TAG}\""
```

File: scripts/docker/test/build-image_arm64.sh [hunk 13]
```diff
@@ -0,0 +1,16 @@
+set -e
+
+arch=arm64
+
+image=arangodb/test-ubuntu
+docker build  --platform linux/$arch -t $image:latest-$arch --file arangodb-test.Dockerfile .
+
+OS_VERSION=$(docker run --rm -it $image:latest-$arch cat etc/os-release | grep -Po "VERSION_ID=\"\K\d+\.\d+")
```
Comment (jsteemann) [inline scripts/docker/test/build-image_arm64.sh:8]: Shouldn't this be an absolute path to `/etc` here?
```suggestion
OS_VERSION=$(docker run --rm -it $image:latest-$arch cat /etc/os-release | grep -Po "VERSION_ID=\"\K\d+\.\d+")
```

File: scripts/docker/test/build-image_x86-64.sh [hunk 14]
```diff
@@ -0,0 +1,16 @@
+set -e
+
+arch=amd64
+
+image=arangodb/test-ubuntu
+docker build  --platform linux/$arch -t $image:latest-$arch --file arangodb-test.Dockerfile .
+
+OS_VERSION=$(docker run --rm -it $image:latest-$arch cat etc/os-release | grep -Po "VERSION_ID=\"\K\d+\.\d+")
```
Comment (jsteemann) [inline scripts/docker/test/build-image_x86-64.sh:8]: Shouldn't this be an absolute path to `/etc` here?
```suggestion
OS_VERSION=$(docker run --rm -it $image:latest-$arch cat etc/os-release | grep -Po "VERSION_ID=\"\K\d+\.\d+")
```

**Review by jsteemann — APPROVED** [review]
I am happy with the RocksDB changes in this PR.
I have left some comments on the shell scripts in this PR, but I don't really know much about them. So I would appreciate if someone else could review these files. Thanks!

**Comment by KVS85** [comment]
Only known (sporadic and fixed in https://github.com/arangodb/arangodb/pull/19733) failures in CI.

**Pull request merged**